Add support for listening on Unix socket

messense opened the pull request:
TODO:

- [x] Safely delete existing socket on startup
- [x] Emit error when both `address` and `port` are specified(But what if I want to listen on both tcp and unix sockets?)
- [x] TLS over unix socket

Closes #545

messense commented:
I am using the `hello_world` example to test it, added a `Rocket.toml` file with `address` set to `unix:///tmp/hello.sock` in `development` section, run `cargo run` to start the server, then use `curl` to send a simple request to it:

```bash
$ cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.0 secs
     Running `/Users/messense/Projects/Rocket/target/debug/hello_world`
🔧  Configured for development.
    => address: unix:///tmp/hello.sock
    => port: 8000
    => log: normal
    => workers: 1
    => secret key: generated
    => limits: forms = 32KiB, json* = 1MiB, msgpack* = 1MiB
    => keep-alive: 5s
    => tls: disabled
    => [extra] hi: "Hello!"
    => [extra] is_extra: true
🛰  Mounting '/':
    => GET / (hello)
🚀  Rocket has launched from http://unix:///tmp/hello.sock:0
```

```bash
$ curl --unix-socket /tmp/hello.sock http://localhost// -v
*   Trying /tmp/hello.sock...
* Connected to localhost (/tmp/hello.sock) port 80 (#0)
> GET // HTTP/1.1
> Host: localhost
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 200 OK
< Content-Type: text/plain; charset=utf-8
< Server: Rocket
< Content-Length: 13
< Date: Wed, 28 Mar 2018 05:40:28 GMT
<
* Connection #0 to host localhost left intact
Hello, world!
```

messense commented on lib/src/data/net_stream.rs:
```diff
@@ -16,6 +17,8 @@ pub enum NetStream {
     #[cfg(feature = "tls")]
     Https(HttpsStream),
+    #[cfg(unix)]
+    UnixHttp(UnixSocketStream),
```
I don't like this, but it is needed to make [concrete_stream](https://github.com/SergioBenitez/Rocket/pull/594/files#diff-58c10788d95c4f599ead66631ae49f04R110) work.

SergioBenitez replied:
Yes, it's all very frustrating.

SergioBenitez commented on lib/src/rocket.rs:
```diff
@@ -91,8 +93,18 @@ impl hyper::Handler for Rocket {
+            #[cfg(unix)]
+            let $server = UnixSocketServer::new(path);
+            #[cfg(not(unix))]
+            panic!("Unix domain socket only supported on Unix platforms");
```
This should be a configuration error.

SergioBenitez commented on lib/src/rocket.rs:
```diff
@@ -91,8 +93,18 @@ impl hyper::Handler for Rocket {
-        $continue
+        let $proto = "http://";
+        if $addr.starts_with("unix://") {
+            let path = &$addr[7..];
```
This should be determined at configuration parsing time. See my comments in the config module.

SergioBenitez commented on lib/src/config/config.rs:
```diff
@@ -379,7 +379,7 @@ impl Config {
     pub fn set_address<A: Into<String>>(&mut self, address: A) -> Result<()> {
         let address = address.into();
-        if (&*address, 0u16).to_socket_addrs().is_err() {
+        if !address.starts_with("unix://") && (&*address, 0u16).to_socket_addrs().is_err() {
```
We also need to validate the path here and check if this is being done on Windows. All configuration validation needs to happen at this point. My guess is that we really want to store the address as an `enum` with variants for TCP, Unix sockets, and even UDP, if it's trivial to add. Something like:

```rust
enum Address {
    Tcp(SocketAddr),
    Udp(SocketAddr),
    /// Stores the validated path to the unix domain socket.
    Unix(PathBuf)
}
```

If there's a better type to use than `PathBuf`, even better.

SergioBenitez commented on lib/src/data/data.rs:
```diff
@@ -8,6 +8,7 @@ use std::time::Duration;
 use super::data_stream::{DataStream, kill_stream};
 use super::net_stream::NetStream;
 use ext::ReadExt;
+#[cfg(unix)] use unix_socket::UnixSocketStream;
```
Putting `cfg(unix)` everywhere seems a bit unfortunate. Can we maybe put all of this functionality (or most of it) in a single module and expose exactly what we need?

SergioBenitez commented on lib/src/data/data.rs:
```diff
@@ -109,6 +110,14 @@ impl Data {
         fn concrete_stream(stream: &&mut NetworkStream) -> Option<NetStream> {
             stream.downcast_ref::<HttpStream>()
                 .map(|s| NetStream::Http(s.clone()))
+                .or_else(|| {
```
This needs to happen whether TLS is or isn't enabled. If you move all of this functionality to a common `unix` module and expose (mostly) no-op shims on non-Unix platforms, this becomes:

```rust
stream.downcast_ref::<HttpStream>()
    .map(|s| NetStream::Http(s.clone()))
    .or_else(try_unix_socket_stream_downcast)
```

SergioBenitez commented on lib/src/rocket.rs:
```diff
@@ -665,7 +677,11 @@ impl Rocket {
         self.fairings.pretty_print_counts();
 
-        let full_addr = format!("{}:{}", self.config.address, self.config.port);
+        let full_addr = if self.config.address.starts_with("unix://") {
```
This should probably be a method on `config`: `config.full_address()`.

SergioBenitez commented on lib/src/unix_socket.rs:
```diff
@@ -0,0 +1,119 @@
+//! hyperlocal provides [hyper](http://github.com/hyperium/hyper) client and server bindings
```
Is this exactly the code from `hyper_local`?

messense replied:
It's [hyperlocal v0.3.0](https://github.com/softprops/hyperlocal/blob/0.3.0/src/lib.rs) with unused code removed thus removed `rustc-serialize` dependency.

SergioBenitez commented:
Awesome, @messense! This is really great. Excited to get this in. :)

messense commented:
> How does this deal with sockets that don't already exist?

It crashes with `Address already in use` error.

```bash
     Running `/Users/messense/Projects/Rocket/target/debug/hello_world`
🔧  Configured for development.
    => address: unix:///tmp/hello.sock
    => port: 8000
    => log: normal
    => workers: 1
    => secret key: generated
    => limits: forms = 32KiB, json* = 1MiB, msgpack* = 1MiB
    => keep-alive: 5s
    => tls: disabled
    => [extra] hi: "Hello!"
    => [extra] is_extra: true
🛰  Mounting '/':
    => GET / (hello)
Error: Rocket failed to launch due to an I/O error.
thread 'main' panicked at 'Address already in use (os error 48)', lib/src/error.rs:207:17
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

messense commented:
Thanks of the detailed review, I hope to address them soon.

messense commented on lib/src/config/config.rs:
```diff
@@ -42,7 +41,7 @@ pub struct Config {
     pub environment: Environment,
     /// The address to serve on.
-    pub address: String,
+    pub address: Address,
```
Change address to `Address` seems make `port` obsolete, should we remove it?

SergioBenitez replied:
We need to keep `port` as a separate property because it's very convenient to be able to change the port independently of the address to listen on, especially via environment variables.

messense commented on lib/src/config/error.rs:
```diff
@@ -48,6 +48,12 @@ pub enum ConfigError {
+    /// The unix socket path already exists
+    ///
+    /// Parameters: (path)
+    BadUnixSocketPath(PathBuf),
```
I am not sure about the naming for `UnsupportedAddress` and `BadUnixSocketPath`.

SergioBenitez replied:
This is good.

messense commented on lib/src/config/builder.rs:
```diff
@@ -89,7 +89,9 @@ impl ConfigBuilder {
-        self.address = address.into();
+        let address = address.into();
+        // FIXME: Remove unwrap
+        self.address = address.parse().unwrap();
```
I'd like to get rid of this `unwrap()` but not sure what's the best way to do it.

SergioBenitez replied:
Keep the address as a `String` in the struct. Errors aren't handled in the individual builders. They're handled at the very end by `finalize()`.

messense commented on lib/src/lib.rs:
```diff
@@ -4,6 +4,7 @@
 #![feature(plugin, decl_macro)]
 #![feature(never_type)]
 #![feature(try_trait)]
+#![feature(transpose_result)]
```
This is need for [`Result<Option<T>>::transpose`](https://doc.rust-lang.org/nightly/std/result/enum.Result.html#method.transpose) method

SergioBenitez replied:
Ideally we wouldn't use this.

messense replied:
Removed its usage.

SergioBenitez commented on lib/src/config/address.rs:
```diff
@@ -0,0 +1,63 @@
+        if s.starts_with("unix://") {
+            let path_str = &s[7..];
+            let path: PathBuf = path_str.into();
+            Ok(Address::Unix(path))
```
This can be one line: `Address::Unix(s[7..].into())`. Better to make these prefixes `const`s though. Something like:

```rust
const UNIX_PREFIX: &str = "unix://";
const TCP_PREFIX: &str = "tcp://";

if string.starts_with(UNIX_PREFIX) {
    Ok(Address::Unix(string[UNIX_PREFIX.len()..].into()))
} else string.starts_with(TCP_PREFIX) {
    ...
} else {
    ...
}
```

SergioBenitez commented on lib/src/config/address.rs:
```diff
@@ -0,0 +1,63 @@
+    }
+}
+
+impl TryFrom<String> for Address {
```
Do we need these `TryFrom` implementations? I don't think they add anything.

SergioBenitez commented on lib/src/config/address.rs:
```diff
@@ -0,0 +1,63 @@
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Address::Tcp(addr) => addr.fmt(f),
+            Address::Unix(ref path) => write!(f, "{}", path.to_string_lossy()),
```
Use `path.display()`.

SergioBenitez commented on lib/src/config/config.rs:
```diff
@@ -240,7 +239,7 @@ impl Config {
                 Config {
                     environment: Development,
-                    address: "localhost".to_string(),
+                    address: "localhost:8000".parse().unwrap(),
```
Just `localhost` here; let's keep the address and port separate.

SergioBenitez commented on lib/src/config/config.rs:
```diff
@@ -377,13 +376,26 @@ impl Config {
-    pub fn set_address<A: Into<String>>(&mut self, address: A) -> Result<()> {
-        let address = address.into();
-        if (&*address, 0u16).to_socket_addrs().is_err() {
+    pub fn set_address<A: TryInto<Address>>(&mut self, address: A) -> Result<()> {
```
This looks really complicated. Why are we using `try_into()` when `parse()` would work just as well?

I'd still like to abstract all of this away into one module with a big `#[cfg(unix)]` and `#[cfg(not(unix))]`. I don't like having these `cfg!` calls sprinkled everywhere; it makes the code much harder to follow.

SergioBenitez commented on lib/src/config/error.rs:
```diff
@@ -101,6 +107,11 @@ impl ConfigError {
+            UnsupportedAddress =>  error!("the address is not supported on this platform"),
+            BadUnixSocketPath(ref path) => {
+                error!("unix socket file path {:?} already exists", path);
+                info_!("Please try again after removing the socket file");
```
Lower case here.

SergioBenitez commented on lib/src/rocket.rs:
```diff
@@ -91,21 +93,40 @@ impl hyper::Handler for Rocket {
+            Address::Unix(path) => {
+                let $proto = "unix://";
+                #[cfg(unix)]
+                let $server = UnixSocketServer::new(path);
```
Will this compile on non-Unix platforms? I think `$continue` uses `$server`, so you'll get a `use before declared` errors.

SergioBenitez commented on lib/src/rocket.rs:
```diff
@@ -91,21 +93,40 @@ impl hyper::Handler for Rocket {
+            },
+            Address::Unix(path) => {
+                let $proto = "unix://";
+                #[cfg(unix)]
```
Can we put this in the `unix` module as well?

messense commented on lib/src/config/config.rs:
```diff
@@ -42,9 +41,9 @@ pub struct Config {
+    pub address: Address,
     /// The port to serve on.
-    pub port: u16,
+    pub port: Option<u16>,
```
To keep `port` separate I have changed `port` from `u16` to `Option<u16>`

messense commented on lib/src/config/config.rs:
```diff
@@ -379,11 +378,28 @@ impl Config {
+                        return Err(ConfigError::BadUnixSocketPath(socket_path.into()));
+                    }
+                    self.address = addr.clone();
+                    self.port = None
```
Here we set `port` to `None` if `address` is an unix socket path.

messense commented on lib/src/config/config.rs:
```diff
@@ -397,13 +413,21 @@ impl Config {
+                self.port = Some(port);
+            }
+            Address::Unix(..) => {
+                return Err(ConfigError::BadPort);
```
`set_port` now returns an `BadPort` error if address was set to unix socket path.

messense commented on lib/src/rocket.rs:
```diff
@@ -90,22 +92,41 @@ impl hyper::Handler for Rocket {
+    ($rocket:expr, $addr:expr, $port:expr, |$server:ident, $proto:ident| $continue:expr) => ({
+        match $addr {
+            Address::Tcp(addr) => {
+                let ($proto, $server) = ("https://", hyper::Server::http((addr.ip(), $port.unwrap())));
```
I kind of feel this defeats the purpose of using `SocketAddr` in `Address:Tcp`

messense commented on lib/src/unix/tls.rs:
```diff
@@ -0,0 +1,183 @@
+use std::io;
```
This module is copied from `hyper-sync-rustls` to workaround [`TlsStream` takes `HttpStream` only as underlying stream impl](https://github.com/SergioBenitez/hyper-sync-rustls/blob/914e770d2697beade61e3a5019e78bcaaf07e8fe/src/lib.rs#L24-L27) now, I'd like to get rid of this and make `hyper_sync_rustls::TlsStream` generic over the underlying stream.

messense commented:
@SergioBenitez Ping for review in case you forgot about this. :-)

kaendfinger commented:
This would be good to have for applications where I might want to use Rocket to implement a local RPC.
:+1:

SergioBenitez commented:
@messense Apologies!

@jebrosen Any chance you'd be able to give this a pass?

jebrosen commented on lib/src/unix/tls.rs:
```diff
@@ -0,0 +1,23 @@
+
```
Stray blank line

jebrosen commented:
Update: `libstd` ensures (with tests) that it does **not** support abstract unix sockets in `UnixListener` and related types, so scratch that. We may want to document this limitation ourselves as well.

messense commented:
I think `libstd` will support abstract unix sockets if someone implement it, here is the tracking issue: https://github.com/rust-lang/rust/issues/42048

SergioBenitez commented:
@messense Argh, I don't think so. Though it shouldn't be too crazy. I plan to make reviewing this PR a priority soon.

What has been implemented in regards to automatic socket removal? Is a socket automatically created by the code in this PR? And, is said socket subsequently removed (presumably not)?

messense commented:
I have done rebasing using `git apply` with manually edited patch file. Although Rocket master currently has a nightly breakage thus I haven't tested it.

> What has been implemented in regards to automatic socket removal? Is a socket automatically created by the code in this PR? And, is said socket subsequently removed (presumably not)?

Current implementation will automatically create socket file if the socket file doesn't exist, and will fail to start if the socket file already exists. Socket removal isn't implemented since there is no clean shutdown method. But I suppose we could provide a configuration for example `auto_socket_removal = true` to automatically try to remove existing socket file on startup.

SergioBenitez commented:
@messense This means that if I run `cargo run; <ctrl-c>; cargo run` with a Unix domain socket, Rocket will fail to start, is that right? That seems really unfortunate.

myfreeweb commented:
Looks like this doesn't provide a way to use an already-open socket (e.g. from [listenfd](https://github.com/mitsuhiko/rust-listenfd))..

juchiast commented:
I would love to use this feature. I use nginx as a proxy to rocket server so using Unix socket will certainly improve the performance.

I see you are currently having problem with reusing the socket. In C I just need to add a `SO_REUSEADDR` flag to the socket, have you try this?

messense commented:
Rebased. Needs some adjustments to the code structure.

messense commented:
@juchiast  With some googling I think `SO_REUSEADDR` does not apply to unix domain socket?

1. [Reusing UNIX domain socket (SO_REUSEADDR for AF_UNIX)](https://gavv.github.io/blog/unix-socket-reuse/)
2. [SO_REUSEADDR and AF_UNIX](https://stackoverflow.com/questions/15716302/so-reuseaddr-and-af-unix)

Would it be OK to add an configuration option to enable/disable deleting the socket file on startup if it exists(Because deleting the file blindly is kind of dangerous.)?

juchiast commented:
Hey @messense, how about using a lock file like [Reusing UNIX domain socket (SO_REUSEADDR for AF_UNIX)](https://gavv.github.io/blog/unix-socket-reuse/) suggested?

messense commented:
@juchiast I have implemented it, see commit 99600370f41106263bb47c6a628f6523cb7d0225

messense commented:
With 99600370f41106263bb47c6a628f6523cb7d0225, when you try to run a Rocket application while another process is using the same socket file, it will fail to launch:

```
    Finished dev [unoptimized + debuginfo] target(s) in 0.20s
     Running `/Users/messense/Projects/Rocket/target/debug/hello_world`
🔧  Configured for development.
    => address: /tmp/hello.sock
    => log: normal
    => workers: 1
    => secret key: generated
    => limits: forms = 32KiB, json* = 1MiB, msgpack* = 1MiB
    => keep-alive: 5s
    => tls: disabled
    => [extra] is_extra: true
    => [extra] hi: "Hello!"
🛰  Mounting '/':
    => GET / (hello)
Error: Rocket failed to launch due to an I/O error.
thread 'main' panicked at 'Resource temporarily unavailable (os error 35)', core/lib/src/error.rs:207:17
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

If no other process is using the same socket file, it will happily delete the existing socket file and launch successfully.

messense commented:
@SergioBenitez @jebrosen I'd love to have a review again and move this forward.

jebrosen commented on core/lib/src/unix.rs:
```diff
@@ -4,11 +4,28 @@ use rocket_http::hyper::net::NetworkStream;
+            return $crate::error::LaunchError::new($crate::error::LaunchErrorKind::Io(err));
+        }
+        if $path.exists() {
+            fs::remove_file($path).unwrap();
```
Same here with returning a `LaunchError`.

Explicitly checking for existence before removal is normally a race condition, but that's probably okay here since it's inside the lock and avoids some extra error finagling afterwards.

jebrosen commented on core/lib/src/unix.rs:
```diff
@@ -0,0 +1,80 @@
+#[cfg(all(unix, feature = "tls"))] use rocket_http::unix::UnixHttpsStream;
+use data::NetStream;
+
+macro_rules! check_unix_domain_socket {
```
Is `lock_unix_domain_socket` a better name for this?

My understanding is that this all works because `lock_file` is still in scope once Rocket yields to hyper at `handle_threads`. I think this should be documented in a comment somewhere.

messense commented on core/lib/src/config/address.rs:
```diff
@@ -0,0 +1,48 @@
+use std::path::PathBuf;
+use std::net::{SocketAddr, ToSocketAddrs};
+
+const UNIX_PREFIX: &str = "unix://";
```
It's `unix://` because we want it to be consistent with `tcp://`, but I think we can also accept `unix:` as an alternative.

messense commented on core/http/src/lib.rs:
```diff
@@ -25,6 +25,7 @@ extern crate state;
 pub mod hyper;
 pub mod uri;
 pub mod ext;
+pub mod unix;
```
Done.

jebrosen commented:
I think I'm happy with the code in its current state. I only have two things I want to verify: that HTTPS works properly over a socket, and that everything works properly on several platforms. I use Linux, IIRC @SergioBenitez uses macOS, and it would be nice to get a confirmation for FreeBSD as well but maybe not as critical.

myfreeweb commented:
Works fine on FreeBSD.

jebrosen commented:
> Works fine on FreeBSD.

Including the lock file check and removal?

myfreeweb commented:
Yes, launching a second instance on the same socket while the first one is running results in "Resource temporarily unavailable", and stopping the first one lets the second one start just fine

messense commented:
I'd love to push this towards the finish line.

jebrosen commented:
I confirmed (with the 'tls' example, via curl) that HTTPS over unix socket works fine on my machine. I think this is ready plus or minus any additional documentation changes.

messense commented on core/lib/src/config/config.rs:
```diff
@@ -383,13 +383,10 @@ impl Config {
                     self.address = addr;
                 }
-                Address::Unix(ref socket_path) => {
+                Address::Unix(..) => {
```
`BadUnixSocketPath` removed in 32fb8caccf79654607c48424904a5b27ee182707.

SergioBenitez commented:
I spent several hours today on code cleanup. Minus documentation, I'm fairly happy with the wrangling I was able to do. That being said, there are four big issues I see here:

  1. With `feature = tls` and TLS disabled, I'm not getting back any body data. This is a bug?
  2. I don't feel comfortable merging this until #801 is resolved. We _need_ to test on non-Unix now.
  3. It's really unfortunate that both the lock file and the socket stick around after the process closes.
  4. We need to give a better error message when locking an existing socket fails. 

I'm okay with merging this without 3, but we should address 1, 2, and 4.

@messense If you add me as a collaborator to your fork, I can push directly to it so you can work off my changes.

SergioBenitez commented:
> I don't have the same problem - I used the tls example, removed the tls configuration, and set address to a unix socket path. curl --unix-socket /path/to/socket http://whatever/ returns Hello, world! as expected.

I tried the `hello_world` example and get nothing back:

```sh
*   Trying rocket.sock...
* Connected to whatever (rocket.sock) port 80 (#0)
> GET / HTTP/1.1
> Host: local
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 200 OK
< Content-Type: text/plain; charset=utf-8
< Server: Rocket
< Content-Length: 13
< Date: Sat, 10 Nov 2018 11:57:37 GMT
<
* Connection #0 to host whatever left intact
```

This happens with my changes and without them. Same thing with the `tls` example. This happens on both macOS and Linux (Arch).

SergioBenitez commented:
> I might be willing to look at #801 tomorrow. Is the obstacle more with Windows in general, or the specifics of Travis's windows environment which I can find no helpful documentation for?

It's really just Windows. The environment that Travis has set up for Windows, the relevant details of which I reference in the post, seems fairly standard. Really, it seems that we need to figure out:

  1. Which additional packages to install, if any.

    I believe we only need to install `sqlite3`. We should really request it from Travis, though, but to get this up and running we should install it with chocolately and build the appropriate `lib` ourselves.

  2. Which environment variables to set to what.

    For each library, it seems we need to set an environment variable to the path where the library lives. I'm not sure which variables these are or where to point them to.

messense commented:
> If you add me as a collaborator to your fork, I can push directly to it so you can work off my changes.

@SergioBenitez I think you can push directly to my branch since GitHub allows edits from maintainers.

![image](https://user-images.githubusercontent.com/1556054/48309987-75f0d880-e5c0-11e8-9cad-3407e8620647.png)

SergioBenitez commented:
@messense Ah, right. Just pushed.

messense commented:
Travis did not pass, seems that the `Address` related changes break something.

messense commented:
> With feature = tls and TLS disabled, I'm not getting back any body data. This is a bug?

I can't reproduce this bug locally.

WIth `tls` enabled:

```bash
$ curl --unix-socket /tmp/tls.sock http://localhost// -v -k
*   Trying /tmp/tls.sock...
* Connected to localhost (/tmp/tls.sock) port 80 (#0)
> GET // HTTP/1.1
> Host: localhost
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 200 OK
< Content-Type: text/plain; charset=utf-8
< Server: Rocket
< Content-Length: 13
< Date: Mon, 12 Nov 2018 02:27:24 GMT
<
* Connection #0 to host localhost left intact
Hello, world!
```

With `tls` disabled:

```bash
$ curl --unix-socket /tmp/tls.sock http://localhost// -v -k
*   Trying /tmp/tls.sock...
* Connected to localhost (/tmp/tls.sock) port 80 (#0)
> GET // HTTP/1.1
> Host: localhost
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 200 OK
< Content-Type: text/plain; charset=utf-8
< Server: Rocket
< Content-Length: 13
< Date: Mon, 12 Nov 2018 02:27:24 GMT
<
* Connection #0 to host localhost left intact
Hello, world!
```

SergioBenitez commented:
@jebrosen @messense Apologies, I was mistaken. It looks like at some point, I updated my shell which changed the way output without terminating new lines was being treated. The prompt was simply eating the output, so while it was there, it wasn't being displayed. As soon as we address issues 2 and 4, this will be ready to merge.

@messense I pushed a fix for that issue.

@jebrosen @messense Is there a need to or precedence for supporting `tcp://`? Similarly, what are the general feelings for `unix://` vs. `unix:` vs. `unix:/`? One thing I want to consider is how easy it is to refer to paths with `.` (current directory). That is, `unix://./foo/bar` vs. `unix:./bar` vs `unix:/./bar/foo`.

messense commented:
@SergioBenitez I have no strong feelings about supporting `tcp://`.

I think nginx and Apache both use `unix:` as unix domain socket prefix.

> The address can be specified as a domain name or IP address, with an optional port, or as a UNIX-domain socket path specified after the “unix:” prefix.

http://nginx.org/en/docs/http/ngx_http_upstream_module.html

> In 2.4.7 and later, support for using a Unix Domain Socket is available by using a target which prepends unix:/path/lis.sock|. For example, to proxy HTTP and target the UDS at /home/www/socket, you would use unix:/home/www.socket|http://localhost/whatever/. Since the socket is local, the hostname used (in this case localhost) is moot, but it is passed as the Host: header value of the request.

https://httpd.apache.org/docs/trunk/mod/mod_proxy.html

Align with them would be intuitive.

messense commented:
> I don't think we need the UnsupportedPort error value anymore

@SergioBenitez Removed in 6d8e9eb

jebrosen commented:
> It looks like at some point, I updated my shell which changed the way output without terminating new lines was being treated.

messense commented:
> We need to give a better error message when locking an existing socket fails.

```bash
$ cargo run
   Compiling rocket v0.4.0-rc.1 (/Users/messense/Projects/Rocket/core/lib)
   Compiling hello_world v0.0.0 (/Users/messense/Projects/Rocket/examples/hello_world)
    Finished dev [unoptimized + debuginfo] target(s) in 6.77s
     Running `/Users/messense/Projects/Rocket/target/debug/hello_world`
🔧  Configured for development.
    => address: unix:///tmp/hello.sock
    => Warning: port: invalid
    => Warning: port is configured, but address is unix domain socket
    => log: normal
    => workers: 1
    => secret key: generated
    => limits: forms = 32KiB, json* = 1MiB, msgpack* = 1MiB
    => keep-alive: 5s
    => tls: disabled
    => [extra] is_extra: true
    => [extra] hi: "Hello!"
🛰  Mounting '/':
    => GET / (hello)
Error: Rocket failed to launch due to lock an existing socket failed:
    => /tmp/hello.lock
    => Note: This usually means that there is another Rocket instance running already.
thread 'main' panicked at 'Resource temporarily unavailable (os error 35)', core/lib/src/error.rs:224:17
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

messense commented:
> Similarly, what are the general feelings for unix:// vs. unix: vs. unix:/?

Changed the prefix from `unix://` to `unix:` in 091f44f, with this change `tcp://` feels inconsistent, so I have removed it.

SergioBenitez commented on core/lib/src/unix.rs:
```diff
@@ -0,0 +1,105 @@
+
+        if let Some(tls) = $rocket.config.tls.clone() {
+            let tls = $crate::http::tls::TlsServer::new(tls.certs, tls.key);
+            let ($proto, $server) = ("unix://", UnixSocketServer::https(path, tls));
```
Same for these two.

SergioBenitez commented on core/lib/src/config/config.rs:
```diff
@@ -355,9 +356,14 @@ impl Config {
+            self.bad_type("address", "string", msg)
+        })?;
+
+        if !cfg!(unix) && address.is_unix() {
```
We should move this check in `Address`.

SergioBenitez commented on core/lib/src/config/config.rs:
```diff
@@ -39,9 +40,9 @@ pub struct Config {
     pub environment: Environment,
     /// The address to serve on.
-    pub address: String,
+    pub address: Address,
```
Because this is `pub`, a user can do something like:

```rust
let mut config = Config::new(..);
config.address = Address::from_str("unix:foo");
rocket::custom(config).launch();
```

On Windows, this will reach the `unreachable!()` in the `serve_unix_socket` macro.

To prevent this, (instead removing the `pub`) we should should ensure that `Address` doesn't parse `unix:` on non-Unix. We should also remove the `Address::Unix` variant entirely on non-Unix.

SergioBenitez commented:
Aside from these changes, I'm fairly happy with this. Once these changes are made and #801 is resolved, I'm happy to merge.

messense commented:
FYI, just saw [hyperlocal-windows](https://github.com/Azure/hyperlocal-windows) crate for unix domain sockets on Windows, but it only supports async hyper.

SergioBenitez commented:
@messense Should we rename `unix:` to `sock:`? That way, when we eventually support sockets on Windows, we won't have to introduce a breaking change.

messense commented:
Confirmed that `cargo build -p rocket --features tls` succeed on Windows.

jebrosen commented:
@SergioBenitez - In the interest of getting this closer to merging, would it be acceptable to have a travis CI run on Windows with only databases disabled? That's something I can do pretty easily now.

SergioBenitez commented:
Alright! It's been half a year, but we _finally_ have a Windows CI! I'd love a rebase on the latest version to try to finally get this in.

messense commented:
@SergioBenitez Rebased.

SergioBenitez commented on core/lib/src/config/address.rs:
```diff
@@ -0,0 +1,63 @@
+    /// The IP address to serve over TCP.
+    Ip(IpAddr),
+    /// The path to the unix domain socket.
+    #[cfg(unix)]
```
Do we need to `cfg` this? I can't see a reason to do so, and it seems more error prone to have it `cfg`d.

SergioBenitez commented on core/lib/src/config/address.rs:
```diff
@@ -0,0 +1,63 @@
+}
+
+impl Address {
+    #[cfg(unix)] crate const UNIX_PREFIX: &'static str = "unix:";
```
Same with this `cfg`.

SergioBenitez commented on core/lib/src/config/address.rs:
```diff
@@ -0,0 +1,63 @@
+
+    crate fn is_unix(&self) -> bool {
+        match self {
+            #[cfg(unix)]
```
And this one.

SergioBenitez commented on core/lib/src/config/address.rs:
```diff
@@ -0,0 +1,63 @@
+    type Err = io::Error;
+
+    fn from_str(string: &str) -> io::Result<Self> {
+        #[cfg(unix)]
```
This one we absolutely need, of course.

SergioBenitez commented on core/lib/src/config/address.rs:
```diff
@@ -0,0 +1,63 @@
+        match self {
+            Address::Hostname(name) => name.fmt(f),
+            Address::Ip(addr) => addr.fmt(f),
+            #[cfg(unix)]
```
But we don't need this one if we remove the `cfg` on the `enum`.

SergioBenitez commented on core/lib/src/config/builder.rs:
```diff
@@ -77,7 +77,7 @@ impl ConfigBuilder {
     ///     .unwrap();
     ///
-    /// assert_eq!(config.address.as_str(), "127.0.0.1");
+    /// assert_eq!(config.address.to_string(), "127.0.0.1");
```
This means this is a breaking change.

SergioBenitez commented on core/lib/src/config/config.rs:
```diff
@@ -340,8 +340,8 @@ impl Config {
-    /// If `address` is not a valid IP address or hostname, returns a `BadType`
-    /// error.
+    /// If `address` is not a valid IP address, hostname or (on Unix platforms
+    /// only) a unix domain socket, returns a `BadType` error.
```
```suggestion
    /// only) a Unix domain socket, returns a `BadType` error.
```

SergioBenitez commented on core/lib/src/config/config.rs:
```diff
@@ -354,10 +354,11 @@ impl Config {
-            return Err(self.bad_type("address", "string", "a valid hostname or IP"));
-        }
+        let address = address.into().parse::<Address>().map_err(|_| {
+            #[cfg(unix)] let msg = "a valid hostname, IP, or unix domain socket path";
```
```suggestion
            #[cfg(unix)] let msg = "a valid hostname, IP, or Unix domain socket path";
```

SergioBenitez commented on core/lib/src/config/config.rs:
```diff
@@ -370,13 +371,17 @@ impl Config {
     #[inline]
-    pub fn set_port(&mut self, port: u16) {
-        self.port = port;
+    pub fn set_port(&mut self, port: u16) -> Result<()> {
```
Why does this return a `Result`?

SergioBenitez commented on core/lib/src/config/config.rs:
```diff
@@ -912,14 +917,24 @@ impl Config {
+        match self.address {
+            Address::Hostname(ref host) => format!("{}:{}", host, self.port),
+            Address::Ip(ip) => format!("{}:{}", ip, self.port),
+            #[cfg(unix)]
```
We shouldn't need this.

SergioBenitez commented on core/lib/src/config/custom_values.rs:
```diff
@@ -42,6 +43,47 @@ impl fmt::Display for SecretKey {
+impl Port {
+    crate fn value(&self) -> u16 {
+        match self {
+            Port::Generated(v) | Port::Provided(v) => *v
```
We shouldn't need the `*v`. Instead, take `self` by value (`self` instead of `&self`). This is just as cheap as what's happening now.

SergioBenitez commented on core/lib/src/config/custom_values.rs:
```diff
@@ -42,6 +43,47 @@ impl fmt::Display for SecretKey {
     }
 }
 
+#[derive(Copy, Clone)]
```
Also derive `Debug`. We should also implement `PartialEq` with `self.value() == other.value()` as the implementation.

SergioBenitez commented on core/lib/src/data/mod.rs:
```diff
@@ -8,3 +8,4 @@ mod from_data;
 pub use self::data::Data;
 pub use self::data_stream::DataStream;
 pub use self::from_data::{FromData, FromDataSimple, Outcome, Transform, Transformed};
+crate use self::net_stream::NetStream;
```
What makes use of this new import?

messense replied:
It's used in `try_socket_stream_downcast`

SergioBenitez commented on core/lib/src/data/net_stream.rs:
```diff
@@ -24,8 +30,10 @@ impl io::Read for NetStream {
         let res = match *self {
-            Http(ref mut stream) => stream.read(buf),
-            #[cfg(feature = "tls")] Https(ref mut stream) => stream.read(buf),
+            Http(ref mut s) => s.read(buf),
```
I'd prefer to call the inner variable `stream` if it fits in 100 cols.

SergioBenitez commented on core/lib/src/error.rs:
```diff
@@ -26,6 +27,8 @@ pub enum LaunchErrorKind {
     Collision(Vec<(Route, Route)>),
     /// A launch fairing reported an error.
     FailedFairings(Vec<&'static str>),
+    /// Lock an existing socket failed.
```
```suggestion
    /// Failed to take a lock on an existing Unix domain socket.
```

SergioBenitez commented on core/lib/src/error.rs:
```diff
@@ -175,6 +179,7 @@ impl std::error::Error for LaunchError {
             LaunchErrorKind::Io(_) => "an I/O error occurred during launch",
             LaunchErrorKind::Collision(_) => "route collisions were detected",
             LaunchErrorKind::FailedFairings(_) => "a launch fairing reported an error",
+            LaunchErrorKind::FailedSocketLock(..) => "lock an existing socket failed",
```
```suggestion
            LaunchErrorKind::FailedSocketLock(..) => "failed to lock an existing unix domain socket",
```

SergioBenitez commented on core/lib/src/error.rs:
```diff
@@ -212,6 +217,12 @@ impl Drop for LaunchError {
                 panic!("launch fairing failure");
             }
+            LaunchErrorKind::FailedSocketLock(ref path, ref e) => {
+                error!("Rocket failed to launch due to lock an existing socket failed:");
```
```suggestion
                error!("Rocket failed lock an existing Unix domain socket:");
```

SergioBenitez commented on core/lib/src/error.rs:
```diff
@@ -212,6 +217,12 @@ impl Drop for LaunchError {
             }
+            LaunchErrorKind::FailedSocketLock(ref path, ref e) => {
+                error!("Rocket failed to launch due to lock an existing socket failed:");
+                info_!("{}", Paint::white(path.display()));
```
```suggestion
                info_!("{}", Paint::default(path.display()).bold());
```

SergioBenitez commented on core/lib/src/error.rs:
```diff
@@ -212,6 +217,12 @@ impl Drop for LaunchError {
+            LaunchErrorKind::FailedSocketLock(ref path, ref e) => {
+                error!("Rocket failed to launch due to lock an existing socket failed:");
+                info_!("{}", Paint::white(path.display()));
+                info_!("Note: This usually means that there is another Rocket instance running already.");
```
```suggestion
                info_!("Note: Another instance of Rocket may be running on the same socket.");
```

SergioBenitez commented on core/lib/src/lib.rs:
```diff
@@ -126,6 +126,7 @@ pub mod http {
     pub use rocket_http::*;
 }
 
+#[macro_use] mod unix;
```
Where do we use the macros from `macro_use`? And, can we use `unix::foo!()` instead?

messense replied:
Not sure why, but I got this error when not using macro_use:

![image](https://user-images.githubusercontent.com/1556054/60762938-5a30ca00-a09c-11e9-85f9-1fbb87d3e93c.png)

jebrosen replied:
My understanding is that macro_rules macros still don't exist in the defining module - they exist in the crate root, unless re-exported with a `use as`. They also must still be defined before use, and annotated with a `#[macro_use]`.

SergioBenitez commented on core/lib/src/unix.rs:
```diff
@@ -0,0 +1,100 @@
+use crate::data::NetStream;
+use crate::http::hyper::net::NetworkStream;
```
Do we need any of the `cfg(unix)` attributes in here? I believe this entire module is only included when `cfg(unix)`.

SergioBenitez commented on core/lib/src/unix.rs:
```diff
@@ -0,0 +1,100 @@
+        .open(&lock_path)?;
+
+    lock_file
+        .try_lock_exclusive()
```
Style wise, I think the following is better:

```rust
lock_file.try_lock_exclusive()
    .map_err(|e| LaunchError::new(LaunchErrorKind::FailedSocketLock(lock_path, e)))?;
```

SergioBenitez commented on scripts/test.sh:
```diff
@@ -96,7 +96,7 @@ if [ "$1" = "--contrib" ]; then
 elif [ "$1" = "--core" ]; then
   FEATURES=(
-    private-cookies # this is already tested since it's the default feature
+    private-cookies
```
Keep the comment.

messense commented:
@SergioBenitez Addressed most of the review comments.

messense commented:
For Windows UDS support we could consider [uds_windows](https://docs.rs/uds_windows/0.1.3/uds_windows/) crate.

SergioBenitez commented:
@messense According to the internet, error 53 is "the network path was not found". Perhaps the path was incorrect?

SergioBenitez commented on core/lib/src/config/address.rs:
```diff
@@ -0,0 +1,60 @@
+use std::path::PathBuf;
+use std::net::{IpAddr, ToSocketAddrs};
+
+/// An enum corresponding to the address to serve on
```
This needs a docstring. How about:

```suggestion
/// The configured address to serve a Rocket application on.
///
/// Addresses can be specified as either IP addresses or hostnames. On Unix
/// platforms, addresses can also be paths to Unix domain sockets specified as
/// `unix:path/to/socket`.
///
/// When a hostname is specified, it is resolved to the _first available_ IP
/// address for the given hostname. Rocket _does not_ bind to all addresses
/// available for a given hostname. For example, if both `127.0.0.1` and `::1`
/// are specified for `localhost`, the first address available of the two will
/// be used, and the other address will be ignored by Rocket.
///
/// When a Unix domain socket is specified, a lock file named
/// `path/to/socket.lock` is created and locked. If locking fails, application
/// launch is aborted. Additionally, if `path/to/socket` does not exist, it is
/// created. Both the lock file and the socket file are unconditionally deleted
/// upon server termination.
```

SergioBenitez commented on core/lib/src/unix.rs:
```diff
@@ -0,0 +1,67 @@
+use std::fs::{self, File};
+
+#[doc(hidden)]
+pub fn lock_socket<P: AsRef<Path>>(path: P) -> Result<File, LaunchError> {
```
Does this need to be `pub`? Can it be `crate` or just private?

messense replied:
Can't be private because it's used in `serve_unix_socket` macro which is used in`serve` macro.

SergioBenitez replied:
Okay. Let's make it `crate` and remove the `doc(hidden)`.

SergioBenitez commented on core/lib/src/unix.rs:
```diff
@@ -0,0 +1,67 @@
+    use fs2::FileExt;
+
+    let path = path.as_ref();
+    let lock_path = path.with_extension("lock");
```
I'm not sure if this does what we want. If `path` is `foo.bar`, this will convert it to `foo.lock` where as we probably want `foo.bar.lock`.

messense replied:
In practice I think I'd name the Unix socket path `foo.sock`/`foo.bar.sock`, so I think it's fine to just replace the extension to `lock`.

SergioBenitez replied:
I'd much prefer to see `foo.sock.lock`.

messense replied:
I don't have a strong opinion about this, if you want to change it feel free to do so.

SergioBenitez commented on core/lib/src/unix.rs:
```diff
@@ -0,0 +1,67 @@
+    // On Unix, the file remains until the last handle to it is closed. We
+    // return the handle to the user. When it is dropped, the lock will free.
+    if path.exists() {
+        fs::remove_file(path)?;
```
I don't understand; what takes care of removing the lock file? What is keeping a handle to the file at `path` alive (we return a handle to `lock_file`, i.e, `lock_path`, not `path`)? Is the file guaranteed to be deleted even if the user force-terminates the server with, say, `ctrl-c`?

messense replied:
1. The lock file isn't removed,
2. `path` is only used for generating `lock_path`,
3. Removing `path` is for ensuing that Unix socket can listen on that path(it will throw an error if the path exists beforehand).

SergioBenitez replied:
_Can_ we remove the lock file? it seems unfortunate to leave files around.

messense replied:
I don't think we can do that now. Removing the lock file automatically would require some shutdown cleanup process (#180) which also isn't guaranteed if process got killed by SIGKILL

SergioBenitez commented on core/lib/src/data/net_stream.rs:
```diff
@@ -59,36 +71,68 @@ impl NetworkStream for NetStream {
+
+#[inline]
+#[cfg(all(unix, not(feature = "tls")))]
+pub fn try_socket_stream_downcast(stream: &mut dyn NetworkStream) -> Option<NetStream> {
```
Can these three function be a lower privacy level than `pub`?

SergioBenitez commented on core/lib/src/config/custom_values.rs:
```diff
@@ -42,6 +43,53 @@ impl fmt::Display for SecretKey {
     }
 }
 
+#[derive(Copy, Clone, Debug)]
```
This needs a docstring. How about:

```suggestion
/// The configured port to serve a Rocket application on.
#[derive(Copy, Clone, Debug)]
```

We should probably implement `From<u16>` for `Port` with an implementation of `Port::Provided(value)`.

SergioBenitez commented on core/lib/src/config/config.rs:
```diff
@@ -368,15 +369,15 @@ impl Config {
     /// let mut config = Config::new(Environment::Staging);
     /// config.set_port(1024);
-    /// assert_eq!(config.port, 1024);
+    /// assert_eq!(config.port, Port::Provided(1024));
```
We shouldn't advocate exposing `Port` like this.

SergioBenitez replied:
```suggestion
    /// assert_eq!(*config.port, 1024);
```

SergioBenitez commented on core/lib/src/config/config.rs:
````diff
@@ -368,15 +369,15 @@ impl Config {
     ///
     /// ```rust
-    /// use rocket::config::{Config, Environment};
+    /// use rocket::config::{Config, Environment, Port};
````
```suggestion
    /// use rocket::config::{Config, Environment};
```

SergioBenitez commented on core/lib/src/config/builder.rs:
```diff
@@ -314,8 +314,9 @@ impl ConfigBuilder {
+
         config.set_address(self.address)?;
-        config.set_port(self.port);
+        config.set_port(self.port.value());
```
```suggestion
        config.set_port(*self.port);
```

(Assuming that works...)

SergioBenitez commented on core/lib/src/config/config.rs:
```diff
@@ -936,7 +946,7 @@ impl fmt::Debug for Config {
     fn eq(&self, other: &Config) -> bool {
         self.address == other.address
-            && self.port == other.port
+            && *self.port == *other.port
```
I think we can now do `self.port == other.port`.

messense commented:
@SergioBenitez 
> According to the internet, error 53 is "the network path was not found". Perhaps the path was incorrect?

Yes, I was using `unix://C:\ProgramData\hello.sock`, it was wrong and should be `unix:C:\ProgramData\hello.sock`.

Got me wronder if we should also support `unix://` prefix, `unix:C:\xxx` feels a little wired.

messense commented:
Unix domain socket support for Windows is added in https://github.com/messense/Rocket/commit/47c9dea55e7817f27ce86952d8638bc72e9a4389

Tested on Windows 10 version 1903

SergioBenitez commented on core/http/src/lib.rs:
```diff
@@ -20,6 +20,9 @@ pub mod hyper;
 pub mod ext;
 
+#[doc(hidden)]
+pub mod unix;
```
I think we still need `cfg(any(unix, windows))` here and everywhere else since `cfg(unix)` + `cfg(windows)` is not every platform.

messense replied:
Fixed.

SergioBenitez commented:
What happens if you try to use UDS on a unsupported version of Windows? Does Rocket still compile on unsupported Windows versions? Is it simple a runtime error? Which error message do we print if so?

messense commented:
Windows Unix domain socket is created with `WSASocketW` API using `AF_UNIX` address family, `WSASocketW` is an existing Windows API and `AF_UNIX` is just a `int` in `winapi` crate. Rocket should compile fine on unsupported Windows versions but return [`WSAEAFNOSUPPORT`](https://docs.microsoft.com/windows/desktop/WinSock/windows-sockets-error-codes-2) meaning that `The specified address family is not supported`.

messense commented:
Side note: `curl` on Windows seems not supporting unix socket yet, but you can use `curl` in WSL to access a Windows Unix socket for example: `curl --unix-socket /mnt/c/ProgramData/hello.sock http://localhost// -v`. Really cool.

messense commented:
I've created a Windows Server 2016 on Azure which does not support UDS, it failed to launch with error message:

![image](https://user-images.githubusercontent.com/1556054/61168088-9bcce380-a57a-11e9-8353-141bd16cd836.png)

```
Error: Rocket failed to bind network socket to given address/port.
thread 'main' panicked at 'An address incompatible with the requested protocol was used. (os error 10047)', core\lib\src\error.rs:197:17
```

What do you think? @SergioBenitez @jebrosen

jebrosen commented:
That looks reasonable. We could detect that error and print a custom message but it might not be worth the effort. We can note in the documentation that this only works on Windows 1809 or later.

messense commented:
> We could detect that error and print a custom message but it might not be worth the effort. 

Added the detection code here: https://github.com/SergioBenitez/Rocket/pull/594/commits/3c9e579007829c42587446ca4ce75d716076081f

cyberb commented:
Do you plan to merge this?

jebrosen commented:
I need to review this again, but I'm generally in favor of this. If it's backwards-compatible we could do it in 0.4, but I suspect the changes to `Address` aren't.

For 0.5 the socket and connection code will need to be rewritten, but it can probably be done somewhat more simply. Furthermore, I would like to make the connection handling extensible such that one could implement unix socket handling, FastCGI, and other I/O all without modifying rocket.

luciusmagn commented:
Hello, what's the status on this? I'd love to use Rocket over unix socket.

Thank you

kotovalexarian commented:
> Furthermore, I would like to make the connection handling extensible such that one could implement unix socket handling, FastCGI, and other I/O all without modifying rocket.

In Ruby ecosystem this is done by separating common HTTP server interface ([Rack](https://github.com/rack/rack)), HTTP server implementations ([Puma](https://puma.io), [Unicorn](https://yhbt.net/unicorn), etc) and HTTP applications ([Ruby on Rails](https://rubyonrails.org), [Sinatra](https://sinatrarb.com), etc).

vanackere commented:
@jebrosen , @SergioBenitez : what is the current status and how would you prefer to have unix-socket support implemented ?
I also have the need for unix socket support for Rocket and wouldn't mind helping if possible...

jebrosen commented:
<https://github.com/SergioBenitez/Rocket/pull/594#issuecomment-535771753> is still accurate. Unix socket support requires roughly these steps:

* The configuration system needs to support specifying a unix socket address. In `master`, the configuration system has been revamped so this will need some tweaks.
* The I/O layer needs to support it. In 0.5, this should now only require an `impl Listener for tokio::net::UnixListener` similar to `impl Listener for tokio::net::TcpListener` (<https://github.com/SergioBenitez/Rocket/blob/master/core/http/src/listener.rs#L164-L180>).
* When the configured address is a unix socket, rocket must set it up (e.g. deleting a stale socket file) and start listening. This can work largely the same as in this PR, but some of the files have been renamed or moved.

At this point I don't believe there are any more significant changes planned for those areas, and I'd be happy to work on and/or review this soon.

@SergioBenitez, any preferences for timing this breaking change to configuration or handling it in an non-breaking way? As far as I can see, the other two changes could be done separately and even in a minor release.

cyberb commented:
I know open-source people hate this, but can I offer $50 (I do not have a pound sterling symbol on my keyboard) for this?

SergioBenitez commented:
This is now wholly outdated. On the one hand, prioritizing this earlier would have prevented staleness. On the other, there are still unresolved questions, and the implementation now would look nothing like the implementation then.

Nevertheless, I've implemented support for a Unix domain socket listener in a local domain branch. The holdback is configuration: how do we configure which listener to use?

There's #1070, which if implemented, would let us say "well, do so programatically." But this isn't particularly satisfying. Starting a Rocket application that listens on a UDS should be as simple as:

```bash
ROCKET_ADDRESS="unix:/path/to/socket" cargo run
```

This is nice and tenable, but the problem is that it ignores the existence of the `port` configuration parameter. What happens to it if the address is a UDS? The approaches to consider are:

  1. Ignore it. Allows setting a `port`, but if the address type doesn't demand a part, simply don't use it.
  2. Error. If a non-default `port` is provided for an address type that doesn't make use of it, emit an error. This seems difficult to implement correctly.
  3. Remove `port` entirely and make it a part of `address`. So what is today `ROCKET_PORT=9000 cargo run` would become `ROCKET_ADDRESS=127.0.0.1:9000 cargo run`. This feels like the "most correct" solution but 1) will almost certainly break almost every deployed Rocket application, and 2) makes changing _just_ the port impossible.

I'm inclined to suggest we follow 1.

In any case, I'm closing out this particular PR as the implementation here has no path forward given the divergence upstream.
